packages/wonder-blocks-data/src/hooks/use-data-internal.js

```diff
@@ -0,0 +1,113 @@
+// @flow
+import {Server} from "@khanacademy/wonder-blocks-core";
+import {useState, useEffect, useContext, useRef} from "react";
+import {RequestFulfillment} from "../util/request-fulfillment.js";
+import {TrackerContext} from "../util/request-tracking.js";
+import {resultFromCacheEntry} from "../util/result-from-cache-entry.js";
+import {ResponseCache} from "../util/response-cache.js";
+
+import type {
+    Result,
+    IRequestHandler,
+    ValidData,
+    CacheEntry,
+} from "../util/types.js";
+
+export const useDataInternal = <TOptions, TData: ValidData>(
+    handler: IRequestHandler<TOptions, TData>,
+    options: TOptions,
+): Result<TData> => {
+    // If we're server-side or hydrating, we'll have a cached entry to use.
+    // So we get that and use it to initialize our state.
+    // This works in both hydration and SSR because the very first call to
+    // this will have cached data in those cases as it will be present on the
+    // initial render - and subsequent renders on the client it will be null.
+    const cachedResult = ResponseCache.Default.getEntry<TOptions, TData>(
+        handler,
+        options,
+    );
+    const [result, setResult] = useState<?CacheEntry<TData>>(cachedResult);
+
+    // We only track data requests when we are server-side and we don't
+    // already have a result, as given by the cachedData (which is also the
+    // initial value for the result state).
+    const maybeTrack = useContext(TrackerContext);
+    if (result == null && Server.isServerSide()) {
+        maybeTrack?.(handler, options);
+    }
+
+    // We need to update our request when the handler changes or the key
+    // to the options change, so we keep track of those.
+    // However, even if we are hydrating from cache, we still need to make the
+    // request at least once, so we do not initialize these references.
+    const handlerRef = useRef();
+    const keyRef = useRef();
+
+    // This effect will ensure that we fulfill the request as desired.
+    useEffect(() => {
+        // If we are server-side, then just skip the effect. We track requests
+        // during SSR and fulfill them outside of the React render cycle.
+        // NOTE: This shouldn't happen since effects would not run on the server
+        // but let's be defensive - I think it makes the code clearer.
+        if (Server.isServerSide()) {
+            return;
+        }
+
+        // If the handler hasn't changed and the key that the options represent
+        // hasn't changed, then we don't need to make our request.
+        if (
+            handler === handlerRef.current &&
```

Comment:
It depends on what triggers the effect to run again. If it is something in this check, then yes, likely not needed, but if it were something else (like the cachedResult), it would be. Since it would be easy to introduce a problem there, I felt being explicit made the code more robust and easier to grok.